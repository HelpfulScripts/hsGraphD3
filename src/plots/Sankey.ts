/**
 * # Sankey Plot
 * 
 * plots Sankley lines. 
 * 
 * ## Usage
 * `graph.add('sankey', {keys:[<node-col>, ...], value:<value-col>, color:<color-scheme>});`
 * - `keys` the columns to use as nodes. At least 2 keys are required. The sequence of keys determines the sequence of the graph flow.
 * - `value` is the {@link SeriesPlot.ValueDef value definition}.
 * - `color` is the color scheme to use, e.g. 'cat10'
 * 
 * 
 * ## Example
 * <example height=200px libs={hsGraphD3:'hsGraphD3'}>
 * <file name='script.js'>
 * const data = {
 *    colNames:['from', 'via', 'to', 'count'], 
 *    rows:[['a', 'sheep', 'one', 45], ['b', 'wolf', 'one', 15], ['c', 'sheep', 'two', 22],
 *          ['c', 'sheep', 'one', 40], ['c', 'wolf', 'three', 23], ['a', 'fence', 'three', 5]
 *    ]
 * };
 * const graph = new hsGraphD3.Graph(root);
 * graph.add('sankey', {keys:['from', 'via', 'to'], value:'count', color:'cat10'});
 * graph.render(data);
 * </file>
 * </example>
 * 
 * ## Example 2: Left aligned
 * <example height=200px libs={hsGraphD3:'hsGraphD3'}>
 * <file name='script.js'>
 * const data = {
 *    colNames:['from', 'to', 'count'], 
 *    rows:[['a', 'sheep', 45], ['sheep', 'one',   45],
 *          ['b', 'one',  15], 
 *          ['c', 'sheep', 22], ['sheep', 'two',   22],
 *          ['c', 'sheep', 40], ['sheep', 'one',   40], 
 *          ['c', 'cow',  23],  ['wolf',  'three', 23], 
 *          ['a', 'fence',  5], ['fence', 'three',  5]
 *    ]
 * };
 * const graph = new hsGraphD3.Graph(root);
 * graph.add('sankey', {keys:['from', 'to'], value:'count', color:'cat10'});
 * graph.defaults.series.series0.align = 'left';
 * graph.render(data);
 * </file>
 * </example>
 * 
 * ## Example 3: Right aligned
 * <example height=200px libs={hsGraphD3:'hsGraphD3'}>
 * <file name='script.js'>
 * const data = {
 *    colNames:['from', 'to', 'count'], 
 *    rows:[['a', 'sheep', 45], ['sheep', 'one',   45],
 *          ['b', 'one',  15], 
 *          ['c', 'sheep', 22], ['sheep', 'two',   22],
 *          ['c', 'sheep', 40], ['sheep', 'one',   40], 
 *          ['c', 'cow',  23],  ['wolf',  'three', 23], 
 *          ['a', 'fence',  5], ['fence', 'three',  5]
 *    ]
 * };
 * const graph = new hsGraphD3.Graph(root);
 * graph.add('sankey', {keys:['from', 'to'], value:'count', color:'cat10'});
 * graph.defaults.series.series0.align = 'right';
 * graph.render(data);
 * </file>
 * </example>
 */

 /** */

import { Log }                  from 'hsutil'; const log = new Log('Sankey');
import { SeriesPlot }           from '../SeriesPlot';
import { CartSeriesDimensions } from '../SeriesPlotCartesian';
import { SeriesPlotDefaults }   from '../SeriesPlot';
import { GraphCfg}              from '../GraphComponent';
import { Series }               from '../Series';
import { DataSet, AccessFn }    from "../Graph";
import { GraphDimensions }      from "../Graph";
import { d3Base }               from "../Settings";
import { sankey as d3Sankey }   from "d3-sankey";
import { SankeyNodeMinimal }    from "d3-sankey";
import { SankeyLinkMinimal }    from "d3-sankey";
import { sankeyLinkHorizontal } from "d3-sankey";
import { sankeyLeft }           from "d3-sankey";
import { sankeyRight }          from "d3-sankey";
import { sankeyCenter }         from "d3-sankey";
import { sankeyJustify }        from "d3-sankey";

Series.register('sankey', (cfg:GraphCfg, sName:string, dims:CartSeriesDimensions) => new Sankey(cfg, sName, dims));
  
export interface SankeyDefaults extends SeriesPlotDefaults {
    width: number;
    padding: number;
    align: 'left' | 'right' | 'center' | 'justified';
}

interface Nodes {
    [index:number]:string,
    names: {}
}


export class Sankey extends SeriesPlot {
    nodes:SankeyNodeMinimal<{}, {}>[];
    links:SankeyLinkMinimal<{}, {}>[];
    sankey:any;
    align = sankeyJustify;

    constructor(cfg:GraphCfg, seriesName:string, dims:CartSeriesDimensions) {
        super(cfg, seriesName, dims);
        this.type = 'none';
    }

    getDefaults(): SankeyDefaults {
        const def = <SankeyDefaults>super.getDefaults();
        def.area.rendered = true;
        def.line.rendered = false;
        def.marker.rendered = true;
        def.label.rendered = false;
        def.width   = 20;
        def.padding = 20;
        def.align   = 'justified'; 
        return def;
    } 

    get dimensions():GraphDimensions { 
        return {};
    }

    public clearStack(data:DataSet) {
    }

    protected updateStack(data:DataSet) {
    }

    public initialize(svg:d3Base, color?:string): void {
        super.initialize(svg, color);

        // if abscissa data is missing, use implicit index as data
        const r = Math.min(this.cfg.viewPort.width, this.cfg.viewPort.height) / 2;
        if (!this.dims.keys)  { log.error(`no keys defined`); }
        if (!this.dims.value)   { log.error(`no value defined`); }
        this.cfg.graph.defaults.scales.margin.bottom = this.cfg.graph.defaults.scales.margin.top;
    }

    public preRender(data:DataSet): void {
        super.preRender(data);
        const margin = this.cfg.graph.defaults.scales.margin;
        switch((<SankeyDefaults>this.defaults).align) {
            case 'left':        this.align = sankeyLeft; break;
            case 'right':       this.align = sankeyRight; break;
            case 'center':      this.align = sankeyCenter; break;
            case 'justified':   this.align = sankeyJustify; break;
        }
        if (this.sankey) {
            this.sankey.update(this.createNodesAndLinks(data))
        } else {
            this.sankey = d3Sankey()
                .nodeId((d:any) => d.name)
                .nodeAlign(this.align)
                .nodeWidth((<SankeyDefaults>this.defaults).width)
                .nodePadding((<SankeyDefaults>this.defaults).padding)
                .extent([[margin.left, margin.top], [this.cfg.viewPort.width - margin.left - margin.right, this.cfg.viewPort.height - margin.top - margin.bottom]]);
            const {nodes, links} = this.sankey(this.createNodesAndLinks(data));
            this.nodes = nodes;
            this.links = links;
        }
    }

    private createNodesAndLinks(data:DataSet):{nodes:Nodes, links:any[]} {
        const keys = (<string[]>this.dims.keys).map(k => data.colNames.indexOf(k));
        const v = data.colNames.indexOf(<string>this.dims.value);
        const nodes:any = this.nodes || [];
        nodes.names = nodes.names || {};
        const links:any = this.links || [];
        links.myKeys = links.myKeys || {};
        data.rows.forEach((r:number[]|string[]) => { 
            for (let i=1; i<keys.length; i++) {
                const s = keys[i-1];
                const d = keys[i];
                if (!nodes.names[r[s]]) { nodes.push(nodes.names[r[s]] = { name: r[s]})}
                if (!nodes.names[r[d]]) { nodes.push(nodes.names[r[d]] = { name: r[d]})}
                const key = `${r[s]}-${r[d]}`;
                if (!links.myKeys[key]) {
                    links.push(links.myKeys[key] = {
                        source: r[s],
                        target: r[d],
                        names:  [r[s], r[d]],
                        value: 0
                    });
                }
                links.myKeys[key].value += r[v];
            }
        });
        return {
            nodes: nodes,
            links: links
        };
    }

    /** renders the component for the given data */
    public renderComponent(data:DataSet): void {
        // opportunity to transform data
        super.renderComponent(data);
    }

    public postRender(data:DataSet): void {
        super.postRender(data);
    }


    //-------------------

    protected d3RenderMarkers(plot:d3Base, data:DataSet) {
        const defaults = this.defaults;
        const popup = this.cfg.components.popup;
        if (defaults.marker.rendered) {
            plot.select('.markers')
                .selectAll("rect")
                .data(this.nodes)
                .join("rect")
                .call(popup.addListener.bind(popup), this.d3RenderPopup(data))
                .transition(this.cfg.transition)
                .call(this.d3DrawMarker.bind(this), data, defaults)
                .call(this.d3MarkerColors.bind(this), data, defaults);
        }
    }

    protected d3RenderFill(plot:d3Base, data:DataSet) {
        const defaults = this.defaults;
        if (defaults.area.rendered) {
            plot.select('.area')
                .attr('fill-opacity', 0)
                .selectAll("g")
                .data(this.links)
                .join("path")
                    .style("mix-blend-mode", "multiply")
                    .attr("d", sankeyLinkHorizontal())
                    .attr("stroke", d => defaults.line.color)
                    .attr("stroke-width", d => Math.max(1, d.width))
                    .call(this.d3FillColors.bind(this), data, defaults)
                    .append("title")
                    .text((d:any) => `${d.source.name} â†’ ${d.target.name}:${d.value}`);
        }
    }

    protected d3RenderLabels(plot:d3Base, data:DataSet):void {
        const defaults = this.defaults;
        const width = this.cfg.viewPort.width;
        if (defaults.label.rendered) {
            plot.select('.label')
                .attr("font-family", "sans-serif")
                .attr("font-size", 10)
                .selectAll("text")
                .data(this.nodes)
                .join("text")
                    .attr("x", d => d.x0 < width / 2 ? d.x1 + 6 : d.x0 - 6)
                    .attr("y", d => (d.y1 + d.y0) / 2)
                    // .attr("dy", "0.35em")
                    .attr("text-anchor", d => d.x0 < width / 2 ? "start" : "end")
                    .text((d:any) => d.name);
            }
    }

    protected d3RenderLine(plot:d3Base, data:DataSet) {
    }

    protected d3DrawMarker(markers:d3Base, data:DataSet, defaults:SeriesPlotDefaults) {
        markers.attr("x", (d:any) => d.x0)
               .attr("y", (d:any) => d.y0)
               .attr("height", (d:any) => d.y1 - d.y0)
               .attr("width", (d:any) => d.x1 - d.x0)
               .attr("fill", defaults.marker.fill.color)
               .attr("opacity", defaults.marker.fill.opacity);
    }

    /**
     * formats the popup string to display
     * @param colNames 
     */
    protected d3RenderPopup(data:DataSet):AccessFn {
        return (r:any) => r.name || '???';
    }
}
  

/*


M149,328
C209,328,209,275,268,275

M536,204
C774,204,774,291,1012,291
*/