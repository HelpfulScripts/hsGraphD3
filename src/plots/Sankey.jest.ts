import * as hsGraphD3   from '../';
import * as d3          from 'd3';
import { SankeyDefaults } from './Sankey';

let clientWidth = 300;
const root = window.document.createElement("div");
root.style.width = `${clientWidth}px`;
root.style.height = "300px";

d3.selection.prototype.transition   = function(){ return this; };
d3.selection.prototype.duration     = function(){ return this; };
d3.selection.prototype.ease         = function(){ return this; };

const data = {
    colNames:['from', 'via', 'to', 'count'], 
    rows:[['a', 'sheep', 'one', 45], ['b', 'wolf', 'one', 15], ['c', 'sheep', 'two', 22],
          ['c', 'sheep', 'one', 40], ['c', 'wolf', 'three', 23], ['a', 'fence', 'three', 5]
    ]
};
   
function createGraph(root:any) {
    const graph = new hsGraphD3.Graph(root);
    graph.isRendered = () => true;
    return graph;
}


describe('Sankey', () => {
    let graph:hsGraphD3.Graph;
    beforeAll(() => {
        graph = createGraph(root);
        graph.add('sankey', {keys:['from', 'via', 'to'], value:'count', color:'cat10'});
        (<SankeyDefaults>graph.defaults.series.series0).align = 'left';
        graph.render(data);
    });
    it('should plot sankey', () => expect(root).toMatchSnapshot());
});
