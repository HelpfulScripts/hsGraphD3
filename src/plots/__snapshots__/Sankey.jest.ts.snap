// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sankey should plot sankey 1`] = `
<div
  style="width: 300px; height: 300px;"
>
  <svg
    class="baseSVG"
    height="100%"
    preserveAspectRatio="xMinYMin meet"
    viewBox="0 0 1000 700"
    width="100%"
  >
    <g
      class="canvas"
    >
      <rect
        class="graphArea"
        fill="#fff"
        fill-opacity="1"
        height="700"
        rx="0"
        ry="0"
        stroke="#00c"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        width="1000"
        x="0"
        y="0"
      />
    </g>
    <g
      class="grids"
    >
      <g
        class="hor-major-Grid"
      />
      <g
        class="hor-minor-Grid"
      />
      <g
        class="ver-major-Grid"
      />
      <g
        class="ver-minor-Grid"
      />
    </g>
    <g
      class="axes"
    >
      <g
        class="horAxis"
      />
      <g
        class="verAxis"
      />
    </g>
    <g
      class="title"
    >
      <text />
    </g>
    <g
      class="series"
    >
      <g
        class="series0"
        style="color: rgb(27, 158, 119);"
      >
        <g
          class="area"
          fill="currentColor"
          fill-opacity="0"
        >
          <path />
          <path
            color="#1f77b4"
            d="M40,110C260,110,260,110,480,110"
            stroke="currentColor"
            stroke-width="180"
            style="mix-blend-mode: multiply;"
          >
            <title>
              a → sheep:45
            </title>
          </path>
          <path
            color="#2ca02c"
            d="M500,190C720,190,720,190,940,190"
            stroke="currentColor"
            stroke-width="340"
            style="mix-blend-mode: multiply;"
          >
            <title>
              sheep → one:85
            </title>
          </path>
          <path
            color="#d62728"
            d="M40,629.9999999999999C260,629.9999999999999,260,630,480,630"
            stroke="currentColor"
            stroke-width="60"
            style="mix-blend-mode: multiply;"
          >
            <title>
              b → wolf:15
            </title>
          </path>
          <path
            color="#8c564b"
            d="M500,538C720,538,720,390,940,390"
            stroke="currentColor"
            stroke-width="60"
            style="mix-blend-mode: multiply;"
          >
            <title>
              wolf → one:15
            </title>
          </path>
          <path
            color="#7f7f7f"
            d="M40,364C260,364,260,324,480,324"
            stroke="currentColor"
            stroke-width="248"
            style="mix-blend-mode: multiply;"
          >
            <title>
              c → sheep:62
            </title>
          </path>
          <path
            color="#bcbd22"
            d="M500,404C720,404,720,484,940,484"
            stroke="currentColor"
            stroke-width="88"
            style="mix-blend-mode: multiply;"
          >
            <title>
              sheep → two:22
            </title>
          </path>
          <path
            d="M40,534C260,534,260,554,480,554"
            stroke="currentColor"
            stroke-width="92"
            style="mix-blend-mode: multiply;"
          >
            <title>
              c → wolf:23
            </title>
          </path>
          <path
            d="M500,614C720,614,720,614,940,614"
            stroke="currentColor"
            stroke-width="92"
            style="mix-blend-mode: multiply;"
          >
            <title>
              wolf → three:23
            </title>
          </path>
          <path
            d="M40,210C260,210,260,478.00000000000006,480,478.00000000000006"
            stroke="currentColor"
            stroke-width="20"
            style="mix-blend-mode: multiply;"
          >
            <title>
              a → fence:5
            </title>
          </path>
          <path
            d="M500,478.00000000000006C720,478.00000000000006,720,558,940,558"
            stroke="currentColor"
            stroke-width="20"
            style="mix-blend-mode: multiply;"
          >
            <title>
              fence → three:5
            </title>
          </path>
        </g>
        <g
          class="markers"
          fill="currentColor"
          fill-opacity="0.75"
          stroke="currentColor"
          stroke-dasharray="none"
          stroke-opacity="1"
          stroke-width="4"
        >
          <rect
            class="popup"
            color="#1f77b4"
            fill="currentColor"
            height="200.00000000000003"
            opacity="0.75"
            width="20"
            x="20"
            y="20"
          />
          <rect
            class="popup"
            color="#2ca02c"
            fill="currentColor"
            height="428.00000000000006"
            opacity="0.75"
            width="20"
            x="480"
            y="20"
          />
          <rect
            class="popup"
            color="#d62728"
            fill="currentColor"
            height="400"
            opacity="0.75"
            width="20"
            x="940"
            y="20"
          />
          <rect
            class="popup"
            color="#8c564b"
            fill="currentColor"
            height="60.000000000000114"
            opacity="0.75"
            width="20"
            x="20"
            y="599.9999999999999"
          />
          <rect
            class="popup"
            color="#7f7f7f"
            fill="currentColor"
            height="151.99999999999994"
            opacity="0.75"
            width="20"
            x="480"
            y="508.00000000000006"
          />
          <rect
            class="popup"
            color="#bcbd22"
            fill="currentColor"
            height="339.9999999999999"
            opacity="0.75"
            width="20"
            x="20"
            y="240.00000000000003"
          />
          <rect
            class="popup"
            fill="currentColor"
            height="88"
            opacity="0.75"
            width="20"
            x="940"
            y="440"
          />
          <rect
            class="popup"
            fill="currentColor"
            height="112.00000000000011"
            opacity="0.75"
            width="20"
            x="940"
            y="548"
          />
          <rect
            class="popup"
            fill="currentColor"
            height="20"
            opacity="0.75"
            width="20"
            x="480"
            y="468.00000000000006"
          />
        </g>
      </g>
    </g>
    <g
      class="popup"
    />
    <style>
      .popup:hover { filter: brightness(0.8);}
    </style>
  </svg>
</div>
`;
